### PR TITLE
Added logic to wait before retrying requests failed with recoverable error

### PR DESCRIPTION
**Pull request by addb**
## Description
- Added logic to wait 5 seconds before retrying request failed with recoverable error
- Refactored code

## Screenshots (if appropriate):
Screenshot 2023-07-28 at 1 02 06 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by yangyansong-adbe — APPROVED** [review]
LGTM

File: code/main/edge/edgeRequestWorker.brs [hunk 1]
```diff
@@ -15,6 +15,9 @@
 
 function _adb_EdgeRequestWorker() as object
     instance = {
+        _RETRY_WAIT_TIME: 5000, ' 5 seconds
```
[the remark this answers is not in the document]
Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:18]: How about doubling everytime?

File: code/main/services/networkResponse.brs [hunk 2]
```diff
@@ -29,18 +29,12 @@ function _adb_NetworkResponse(responseCode as integer, responseBody as string) a
...
+            return m._responseCode >= 200 and m._responseCode < 300
         end function,
 
         isRecoverable: function() as boolean
             ' RECOVERABLE_ERROR_CODES = [408, 504, 503]
-            if m._responseCode = 408 or m._responseCode = 504 or m._responseCode = 503 then
-                return true
-            end if
-            return false
+            return  m._responseCode = 408 or m._responseCode = 504 or m._responseCode = 503
```
Comment (emdobrin) [inline code/main/services/networkResponse.brs:37]: for mobile the recoverable codes are: 408, 429, 502, 503, 504. Can we match them here?

File: code/main/edge/edgeRequestWorker.brs [hunk 3]
```diff
@@ -59,8 +62,9 @@ function _adb_EdgeRequestWorker() as object
             responseArray = []
             while m._queue.count() > 0
 
-                if _adb_isEmptyOrInvalidString(ecid) or _adb_isEmptyOrInvalidString(configId) then
-                    _adb_logWarning("processRequests() - Edge request skipped. ECID and/or configId not set.")
+                currTS = _adb_timestampInMillis()
+                if (m._lastFailedRequestTS <> m._INVALID_WAIT_TIME) and ((currTS - m._lastFailedRequestTS) < m._RETRY_WAIT_TIME)
+                    ' Wait for 5 seconds before retrying the hit failed with recoverable error.
```
Comment (emdobrin) [inline code/main/edge/edgeRequestWorker.brs:67]: nit: Wait for _RETRY_WAIT_TIME_MILLIS before ...

File: code/main/edge/edgeRequestWorker.brs [hunk 4]
```diff
@@ -82,8 +86,11 @@ function _adb_EdgeRequestWorker() as object
                 if networkResponse.isSuccessful()
                     edgeResponse = _adb_EdgeResponse(requestId, networkResponse.getResponseCode(), networkResponse.getResponseString())
                     responseArray.Push(edgeResponse)
+                    ' Request sent out successfully
+                    m._lastFailedRequestTS = m._INVALID_WAIT_TIME
                 else if networkResponse.isRecoverable()
-                    _adb_logError("processRequests() - Edge request failed with recoverable error. Request will be retried.")
+                    m._lastFailedRequestTS = _adb_timestampInMillis()
+                    _adb_logError("processRequests() - Edge request failed with recoverable error:(" + networkResponse.getResponseString() + "). Request will be retried.")
```
Comment (emdobrin) [inline code/main/edge/edgeRequestWorker.brs:93]: _adb_logWarning("processRequests() - Edge request failed with recoverable error code:(" + networkResponse.getResponseCode() + "). Request will be retried after " + StrI(_RETRY_WAIT_TIME_MILLIS) + " ms.")

File: code/main/edge/edgeRequestWorker.brs (hunk 1 above)
[the remark this answers is not in the document]
Comment (emdobrin) [reply code/main/edge/edgeRequestWorker.brs:18]: for online sessions 24h as retry timeout sounds too much, we can max out at something like 5mins.
@praveek are you talking about online sessions or also downloaded or over antenna content playing?

File: code/main/services/networkResponse.brs (hunk 2 above)
Comment (yangyansong-adbe) [reply code/main/services/networkResponse.brs:37]: I copied those recoverable codes from the mobile core : https://github.com/adobe/aepsdk-core-ios/blob/eccca8b8cbc47a41edbc30df5ff6b398fb094091/AEPServices/Sources/network/NetworkServiceConstants.swift#L16

Where is the code to handle other recoverable codes?

Comment (yangyansong-adbe) [reply code/main/services/networkResponse.brs:37]: ```502 error is usually not something you can fix, but requires a fix by the web server or the proxies you are trying to get access through.```

It seems that we are able to include `429`, but not `502`.

**Comment by yangyansong-adbe** [comment]
@addb  The API callback timeout is 5s. If we decide to increase the retry interval beyond 5s, we need to reset the `_lastFailedRequestTS` when we add a new event to the queue, after this line: https://github.com/adobe/aepsdk-roku/blob/78f224670370eb0995168198f764634f6bc36ad0/code/main/edge/edgeRequestWorker.brs#L46C23-L46C23

File: code/main/edge/edgeRequestWorker.brs [hunk 5]
```diff
@@ -59,8 +65,9 @@ function _adb_EdgeRequestWorker() as object
             responseArray = []
             while m.hasQueuedEvent()
 
-                if _adb_isEmptyOrInvalidString(ecid) or _adb_isEmptyOrInvalidString(configId) then
-                    _adb_logWarning("processRequests() - Edge request skipped. ECID and/or configId not set.")
+                currTS = _adb_timestampInMillis()
+                if (m._lastFailedRequestTS <> m._INVALID_WAIT_TIME) and ((currTS - m._lastFailedRequestTS) < m._RETRY_WAIT_TIME_MS)
```
Comment (yangyansong-adbe) [inline code/main/edge/edgeRequestWorker.brs:69]: Optional: we can eliminate the use of `_INVALID_WAIT_TIME` and introduce a new private function, `_resetLatFailedRequestTS()`, which will update `m._lastFailedRequestTS` to any negative value and do `if (m._lastFailedRequestTS > 0 ) and (...)` here.

File: code/tests/test_helper.brs [hunk 6]
```diff
@@ -48,3 +48,10 @@ function isEmptyOrInvalidString(str as dynamic) as boolean
 
     return false
 end function
+
+
+' ************************ Wait Timer Helper ************************
+function _adb_waitFor(time = 1 as integer)
```
Comment (yangyansong-adbe) [inline code/tests/test_helper.brs:54]: We should remove it since it appears to be unused.

**Review by yangyansong-adbe — APPROVED** [review]

File: code/tests/test_helper.brs (hunk 6 above)
Comment (addb) [reply code/tests/test_helper.brs:54]: Yes

File: code/main/edge/edgeRequestWorker.brs (hunk 5 above)
Comment (addb) [reply code/main/edge/edgeRequestWorker.brs:69]: I would prefer to keep the current logic, if you are okay with it.

**Pull request merged**